test/AutoDiff/downstream/reabstraction_e2e.swift

```diff
@@ -0,0 +1,178 @@
+  expectEqual(Float(7 * 3 * 3 + 10), transformed(3, 10))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3) { transformed($0, 10) })
+}
+
+ReabstractionE2ETests.test("diff param and nondiff param generic => concrete") {
+  func inner<T: HasFloat>(x: T, y: T) -> Float {
+    7 * x.float * x.float + y.float
+  }
+  let transformed: @differentiable (Float, @noDerivative Float) -> Float = inner
+  expectEqual(Float(7 * 3 * 3 + 10), transformed(3, 10))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3) { transformed($0, 10) })
+}
+
+ReabstractionE2ETests.test("result generic => concrete") {
+  func inner<T: HasFloat>(x: Float) -> T {
+    T(float: 7 * x * x)
+  }
+  let transformed: @differentiable (Float) -> Float = inner
+  expectEqual(Float(7 * 3 * 3), transformed(3))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3, in: transformed))
+}
+
+ReabstractionE2ETests.test("diff param concrete => generic => concrete") {
+  typealias FnTy<T: Differentiable> = @differentiable (T) -> Float
+  func id<T: Differentiable>(_ f: @escaping FnTy<T>) -> FnTy<T> { f }
+  func inner(_ x: Float) -> Float {
+    7 * x * x
+  }
+  let transformed = id(inner)
+  expectEqual(Float(7 * 3 * 3), transformed(3))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3, in: transformed))
+}
+
+ReabstractionE2ETests.test("nondiff param concrete => generic => concrete") {
+  typealias FnTy<T: Differentiable> = @differentiable (Float, @noDerivative T) -> Float
+  func id<T: Differentiable>(_ f: @escaping FnTy<T>) -> FnTy<T> { f }
+  func inner(_ x: Float, _ y: Float) -> Float {
+    7 * x * x + y
+  }
+  let transformed = id(inner)
+  expectEqual(Float(7 * 3 * 3 + 10), transformed(3, 10))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3) { transformed($0, 10) })
+}
+
+ReabstractionE2ETests.test("diff param and nondiff param concrete => generic => concrete") {
+  typealias FnTy<T: Differentiable> = @differentiable (T, @noDerivative T) -> Float
+  func id<T: Differentiable>(_ f: @escaping FnTy<T>) -> FnTy<T> { f }
+  func inner(_ x: Float, _ y: Float) -> Float {
+    7 * x * x + y
+  }
+  let transformed = id(inner)
+  expectEqual(Float(7 * 3 * 3 + 10), transformed(3, 10))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3) { transformed($0, 10) })
+}
+
+ReabstractionE2ETests.test("result concrete => generic => concrete") {
+  typealias FnTy<T: Differentiable> = @differentiable (Float) -> T
+  func id<T: Differentiable>(_ f: @escaping FnTy<T>) -> FnTy<T> { f }
+  func inner(_ x: Float) -> Float {
+    7 * x * x
+  }
+  let transformed = id(inner)
+  expectEqual(Float(7 * 3 * 3), transformed(3))
+  expectEqual(Float(7 * 2 * 3), gradient(at: 3, in: transformed))
+}
+
+ReabstractionE2ETests.test("@differentiable function => opaque generic => concrete") {
+  func id<T>(_ t: T) -> T { t }
+  let inner: @differentiable (Float) -> Float = { 7 * $0 * $0 }
+  let transformed = id(inner)
+
+  // TODO(TF-1122): Actually using `transformed` causes a segfault at runtime.
```

Comment:
How about adding a dedicated `not --crash` negative test for TF-1122? In addition to commenting out the crashers in this test.

Reply:
done